autonomy/deploy/base.py

```diff
@@ -69,33 +68,33 @@ def __init__(
         keys: Path,
         number_of_agents: Optional[int] = None,
         private_keys_password: Optional[str] = None,
+        agent_instances: Optional[List[str]] = None,
     ) -> None:
         """Initialize the Base Deployment."""
-        self.private_keys: List = []
+        self.keys: List = []
         self.private_keys_password = private_keys_password
         self.service = load_service_config(service_path)
         if number_of_agents is not None:
             self.service.number_of_agents = number_of_agents
 
+        self.agent_instances = agent_instances
         self.read_keys(keys)
-        if self.service.number_of_agents > len(self.private_keys):
-            raise ValueError("Number of agents cannot be greater than available keys.")
 
     def read_keys(self, file_path: Path) -> None:
         """Read in keys from a file on disk."""
 
         keys = json.loads(file_path.read_text(encoding=DEFAULT_ENCODING))
-
-        key_schema = (
-            KEY_SCHEMA_UNENCRYPTED_KEY
-            if self.private_keys_password is None
-            else KEY_SCHEMA_ENCRYPTED_KEY
-        )
         for key in keys:
-            for required_key in [KEY_SCHEMA_ADDRESS, key_schema]:
-                if required_key not in key.keys():
-                    raise ValueError("Key file incorrectly formatted.")
-            self.private_keys.append(key[key_schema])
+            if {KEY_SCHEMA_ADDRESS, KEY_SCHEMA_PRIVATE_KEY} != set(key.keys()):
+                raise ValueError("Key file incorrectly formatted.")
+
+        if self.agent_instances is not None:
+            keys = [kp for kp in keys if kp["address"] in self.agent_instances]
+            self.service.number_of_agents = len(keys)
+
+        self.keys = keys
+        if self.service.number_of_agents > len(self.keys):
+            raise ValueError("Number of agents cannot be greater than available keys.")
```

Comment:
1. 
can an agent have multiple keys? If no, check for this as well, if yes, then this last check might lead to a situation where one agent has multiple and another none. Also now complexity is n * m.

```
agent_addresses = set(self.agent_instances)  
keys = [kp for kp in keys if kp["address"] in agent_addresses]
```

2. 
- what if number_of_agents is not set on `__init__` and no agent instances here? 
`self.service.number_of_agents > len(self.keys)` -> `None > len(self.keys)` -> will error out
- what if `self.service.number_of_agents` is already set to `int(x)` and this doesn't match `len(keys)` in the overwrite here?

Reply:
1. no, an agent can have one key only atm.

Reply:
1. Agents cannot have multiple keys AFAIK
2. We define `number_of_agents ` in service config, then perform an override in `__init__` and `read_keys` method of service specification object. 
3. I don't understand how it won't match, we need the number of agents to be less than or equal to the number of keys.